pints/_mcmc/_relativistic.py

```diff
@@ -158,6 +161,68 @@ def ask(self):
         self._ready_for_tell = True
         return np.array(self._position, copy=True)
 
+    def _calculate_momentum_distribution(self):
+        """Calculate an approximation to the CDF of momentum magnitude.
+
+        The purpose of this method is to calculate self._inv_cdf, a function
+        giving a numerical approximation to the inverse cumulative distribution
+        function of the magnitude of the momentum vector. This function can be
+        used to perform inverse transform sampling to generate samples of the
+        momentum.
+
+        This method should be called before any sampling is performed. As
+        implemented, it is called upon the first call to self.ask.
+
+        Numerical integration of the momentum density is performed using the
+        trapezoidal rule.
+        """
+        num_adaptations = 0
+
+        # Set initial values for the integration grid
+        max_value = 10
+        spacing = 1e-4
+
+        integration_accepted = False
+
+        while not integration_accepted:
+            # Evaluate the logpdf on a grid
+            integration_grid = np.arange(1e-6, max_value, spacing)
+            logpdf_values = self._momentum_logpdf(integration_grid)
+
+            # Integrate using the trapezoidal rule
+            cdf = np.logaddexp.accumulate(
+                np.logaddexp(logpdf_values[1:], logpdf_values[:-1]))
+
+            # Normalize the CDF
+            cdf = np.exp(cdf - cdf[-1])
+
+            # Adapt the integration grid if the result is inaccurate
+            if max(np.diff(cdf)) > 1e-4:
+                # cdf is changing too much, so a finer grid is required
+                spacing /= 2
+                num_adaptations += 1
+
+            elif cdf[-1] - cdf[-2] > 1e-8:
+                # cdf is still increasing at the end
+                max_value *= 2
+                num_adaptations += 1
+
+            else:
+                integration_accepted = True
+
+            if num_adaptations > 6:
```

Comment:
Any reason to adapt 6 times only? Just curious.

Reply:
It's very heuristic, but that's the point at which it started to get really slow and also expensive in memory.
For example, when I increase to 10 adaptions, github doesn't give us enough memory for the test that goes through all of them to pass on windows.
I'll try coming up with a better approach that considers the current size of the integration grid, and puts a reasonable limit on how large it can get.